core/domain/story_services.py

```diff
@@ -187,33 +189,35 @@ def apply_change_list(story_id, change_list):
         raise
 
 
-def _save_story(committer_id, story, commit_message, change_list):
-    """Validates a story and commits it to persistent storage. If
-    successful, increments the version number of the incoming story domain
-    object by 1.
+def validate_explorations(exp_ids, raise_error):
+    """Validates the explorations in the given story and checks whether they
+    are compatible with the mobile app.
 
     Args:
-        committer_id: str. ID of the given committer.
-        story: Story. The story domain object to be saved.
-        commit_message: str. The commit message.
-        change_list: list(StoryChange). List of changes applied to a story.
+        exp_ids: list(str). The exp IDs to validate.
+        raise_error: bool. Whether to raise an Exception when a validation error
+            is encountered. If not, a list of the error messages are
+            returned. raise_error should be True when this is called before
+            saving the story and False when this function is called from the
+            frontend.
+
+    Returns:
+        list(str). The various validation error messages
+            (if raise_error is True).
 
     Raises:
-        ValidationError: An invalid exploration was referenced in the
-            story.
-        Exception: The story model and the incoming story domain
-            object have different version numbers.
+        ValidationError. Expected story to only reference valid explorations.
+        ValidationError. Exploration with ID is not public. Please publish
+            explorations before adding them to a story.
+        ValidationError. All explorations in a story should be of the same
+            category.
+        ValidationError. Invalid language found for exploration.
+        ValidationError. Expected no exploration to have parameter values in it.
+        ValidationError. Invalid interaction in exploration.
+        ValidationError. RTE content in state of exploration with ID is not
+            supported on mobile.
     """
-    if not change_list:
-        raise Exception(
-            'Unexpected error: received an invalid change list when trying to '
-            'save story %s: %s' % (story.id, change_list))
-
-    story.validate()
-    # Validate that all explorations referenced by the story exist.
-    exp_ids = [
-        node.exploration_id for node in story.story_contents.nodes
-        if node.exploration_id is not None]
+    validation_error_messages = []
 
     # The first exp ID in the story to compare categories later on.
     sample_exp_id = exp_ids[0] if exp_ids else None
```

Comment:
If the first exp_id is invalid it won't return all the error logs (like different category etc.) but it will report all the errors if th2 2nd id is invalid, right? is that expected?

Reply:
Oh, yeah, fixed it! sample_exp_id is now found by going through exp_ids and getting the first one in exp_dict.